src/components/KTableData/KTableData.vue

```diff
@@ -0,0 +1,706 @@
+const getHeaderSlots = computed((): TableColumnSlotName[] => {
+  if (!slots) {
+    return []
+  }
+
+  return Object.keys(slots).filter((slot) => slot.startsWith('column-')) as TableColumnSlotName[]
+})
+
+const getHeaderTooltipSlots = computed((): TableColumnTooltipSlotName[] => {
+  if (!slots) {
+    return []
+  }
+
+  return Object.keys(slots).filter((slot) => slot.startsWith('tooltip-')) as TableColumnTooltipSlotName[]
+})
+
+const getCellSlots = computed((): string[] => {
+  if (!slots) {
+    return []
+  }
+
+  return Object.keys(slots).filter((slot) => tableHeaders.value.some((header) => header.key === slot))
+})
+
+const isInitialFetch = ref<boolean>(true)
+const fetchData = async () => {
+  const searchInput = props.searchInput
+
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore
+  const res = await props.fetcher({
+    pageSize: pageSize.value,
+    page: page.value,
+    query: searchInput || filterQuery.value,
+    sortColumnKey: sortColumnKey.value,
+    sortColumnOrder: sortColumnOrder.value,
+    offset: offset.value,
+  })
+  tableData.value = res.data as Record<string, any>[]
+  total.value = props.paginationAttributes?.totalCount || res.total || res.data?.length
+
+  if (props.paginationAttributes?.offset) {
+    if (!res.pagination?.offset) {
+      nextOffset.value = null
+    } else {
+      nextOffset.value = res.pagination.offset
+
+      if (!offsets.value[page.value]) {
+        offsets.value.push(res.pagination.offset)
+      }
+    }
+
+    hasNextPage.value = (res.pagination && 'hasNextPage' in res.pagination) ? res.pagination.hasNextPage : true
+  }
+
+  // if the data is empty and the page is greater than 1,
+  // e.g. user deletes the last item on the last page,
+  // reset the page to 1
+  if (tableData.value.length === 0 && page.value > 1) {
+    page.value = 1
+    offsets.value = [null]
+    offset.value = null
+  }
+
+  isInitialFetch.value = false
+
+  return res
+}
+
+const initData = () => {
+  const fetcherParams = {
+    ...defaultFetcherProps,
+    ...props.initialFetcherParams,
+  }
+  // don't allow overriding default settings with `undefined` values
+  page.value = fetcherParams.page ?? defaultFetcherProps.page
+  pageSize.value = fetcherParams.pageSize ?? defaultFetcherProps.pageSize
+  filterQuery.value = fetcherParams.query ?? defaultFetcherProps.query
+  sortColumnKey.value = fetcherParams.sortColumnKey ?? defaultFetcherProps.sortColumnKey
+  sortColumnOrder.value = fetcherParams.sortColumnOrder as SortColumnOrder ?? defaultFetcherProps.sortColumnOrder as SortColumnOrder
+
+  if (props.clientSort && sortColumnKey.value && sortColumnOrder.value) {
+    defaultClientSideSorter(sortColumnKey.value, '', sortColumnOrder.value, tableData.value)
+  }
+
+  if (props.paginationAttributes?.offset) {
+    offset.value = fetcherParams.offset
+    offsets.value.push(fetcherParams.offset)
+  }
+
+  // trigger setting of tableFetcherCacheKey
+  hasInitialized.value = true
+}
```

Comment:
nit: can you add comments on this block?

Reply:
Done